Std/Data/String/Basic.lean

```diff
@@ -1,17 +1,83 @@
 /-
 Copyright (c) 2022 Jannis Limperg. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Jannis Limperg
+Authors: Jannis Limperg, James Gallicchio, F. G. Dorais
 -/
 
 import Std.Data.Char
 import Std.Data.Nat.Lemmas
+import Std.Data.Array.Match
 
 instance : Coe String Substring := ⟨String.toSubstring⟩
 
-protected theorem String.Pos.ne_zero_of_lt : {a b : Pos} → a < b → b ≠ 0
+namespace String
+
+protected theorem Pos.ne_zero_of_lt : {a b : Pos} → a < b → b ≠ 0
 | _, _, hlt, rfl => Nat.not_lt_zero _ hlt
 
+/-- Knuth-Morris-Pratt matcher type
+
+This type is used to keep data for running the Knuth-Morris-Pratt (KMP) string matching algorithm.
+KMP is a linear time algorithm to locate all substrings of a string that match a given pattern.
+Generating the algorithm data is also linear in the length of the pattern but the data can be
+re-used to match the same pattern over different strings.
+
+The KMP data for a pattern string can be generated using `Matcher.ofString`. Then `Matcher.find?`
+and `Matcher.findAll` can be used to run the algorithm on an input string.
+```
+def m := Matcher.ofString "abba"
+
+#eval Option.isSome <| m.find? "AbbabbA" -- false
+#eval Option.isSome <| m.find? "aabbaa" -- true
+
+#eval Array.size <| m.findAll "abbabba" -- 2
+#eval Array.size <| m.findAll "abbabbabba" -- 3
+```
+-/
+abbrev Matcher := Array.Matcher Char
+
+/-- Make KMP matcher from pattern substring -/
+@[inline] def Matcher.ofSubstring (pattern : Substring) : Matcher :=
+  Array.Matcher.ofStream pattern
+
+/-- Make KMP matcher from pattern string -/
+@[inline] def Matcher.ofString (pattern : String) : Matcher :=
+  Matcher.ofSubstring pattern
+
+/-- The string pattern for the matcher -/
+def Matcher.pattern (m : Matcher) : String :=
+  m.table.foldl (fun s (c,_) => s.push c) ""
+
+/-- The byte size of the string pattern for the matcher -/
+def Matcher.patternSize (m : Matcher) : Nat :=
+  String.Pos.byteIdx <| m.table.foldl (fun n (c,_) => n + c) 0
```

Comment:
It is surprising that this is not O(1), and I believe this leads to an unexpected overhead in `findAll` below, which calls it in every iteration. (You shouldn't use `let rec` for precomputed state! It's just another recursive function.)

Reply:
Yes, that's much better!